# Introduction to Data Science (CS4661). 
## Homework4

#### By: Harshil Patel(306646748)
---------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------

### Question1: Cancer Diagnosis Using Machine Learning

In this homework, we work with a real dataset from UCI Database.


In [11]:
#importing needed libraries

import numpy as np
import pandas as pd

# importing KNeighborsClassifier for KNN classification
from sklearn.neighbors import KNeighborsClassifier

#importing spliting method
from sklearn.model_selection import train_test_split

#importing acccuracy_score to check accuracy
from sklearn.metrics import accuracy_score

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier

from sklearn.ensemble import RandomForestClassifier

### a- 
Read the dataset file “Cancer.csv” (from github using the following command), and assign
it to a Pandas DataFrame:
df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Cancer.csv")
Check out the dataset. As you see, the dataset includes 9 numerical features. The last
column is the binary label (“1” means it is a malignant cancer, “0” means it is a benign
tumor). You will use all 9 features in this homework.


In [13]:
import pandas as pd

df= pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Cancer.csv")
df.head()

Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
0                5                        1                         1   
1                5                        4                         4   
2                3                        1                         1   
3                6                        8                         8   
4                4                        1                         1   

   Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
0                  1                            2            1   
1                  5                            7           10   
2                  1                            2            2   
3                  1                            3            4   
4                  3                            2            1   

   Bland_Chromatin  Normal_Nucleoli  Mitoses  Malignant_Cancer  
0                3                1        1                 0  
1                3                2        1                 0  
2                3                1        1                 0  
3                3                7        1                 0  
4                3                1        1                 0

### b- 
Use sklearn functions to split the dataset into testing and training sets with the following
parameters: test_size=0.35, random_state=3.

In [15]:
#importing train_test_split from sklearn.model_selection
from sklearn.model_selection import train_test_split

#making a list of the features that are needed
features = ['Clump_Thickness','Uniformity_of_Cell_Size','Uniformity_of_Cell_Shape','Marginal_Adhesion','Single_Epithelial_Cell_Size','Bare_Nuclei','Bland_Chromatin','Normal_Nucleoli','Mitoses']

#selecting above features from the data
X = data[features]

#selecting labels from the data
y = data['Malignant_Cancer']

#spliting the data randomly into training and testing data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35, random_state=3)

# print the size of the traning set:
print(X_train.shape)
print(y_train.shape)

# print the size of the testing set:
print(X_test.shape)
print(y_test.shape)

(97, 9)
(97,)
(53, 9)
(53,)


### c- 
Use “Decision Tree Classifier” to predict Cancer based on the training/testing datasets
that you built in part (b). Then, calculate and report the accuracy of your classifier. Use
this command to define your tree:
my_DecisionTree = DecisionTreeClassifier(random_state=3).

In [16]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

tree = DecisionTreeClassifier(random_state=3)
tree.fit(X_train, y_train)


#defining decision tree object
my_DecisionTree = DecisionTreeClassifier(random_state=3)

In [17]:
#train data using decision tree object

my_DecisionTree.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=3,
            splitter='best')

In [18]:
#predict test data
y_predict = my_DecisionTree.predict(X_test)

In [19]:
#check accuracy
accuracy_log = accuracy_score(y_test,y_predict)
print(accuracy_log)

0.8301886792452831


### d- 
Now, we want to perform a new Ensemble Learning method called “Bagging” based on
Voting on 19 decision tree classifiers.
Note: you should write your own code to perform Bagging (don’t use scikit-learn functions
for Bagging!)
To do so, you need to perform bootstrapping first. You can write a “for” loop with loop
variable i = 0…18.
In each iteration of the loop, you have to:




1. make a bootstarp sample of the original “Training” Dataset (build in part(b)) with the
size of bootstarp_size = 0.8*(Size of the original dataset). You can use the following
command to generate a random bootstrap dataset (“i" is the variable of the loop, so
the random_state changes in each iteration):
resample(X_train, n_samples = bootstarp_size , random_state=i , replace = True)

2. Define and train a new base decision tree classifier on this dataset in each iteration:
Base_DecisionTree = DecisionTreeClassifier(random_state=3).

3. Perform prediction using “this base classifier” on the original “Testing” Dataset X_test
(build in part(b)), and save the prediction results for all testing samples.
After finishing the “for” loop, you should have 19 different predictions for EACH sample
in your testing set. Then, Perform Voting to make the final decision on each data sample
based on the votes of all 19 classifiers.
Finally, calculate and report the final accuracy of your Bagging (Voting) method.
Note: You do NOT need to calculate the accuracy of each one of the base classifiers in
each round of the loop! You have to just perform Voting to make the final decision on
each data sample, and then calculate the accuracy on the final results.

### e- 
Use scikit-learn “Random Forest” classifier to predict Cancer based on the training/testing
datasets that you built in part (b). Then, calculate and report the accuracy of your
classifier. Use this command to import and define your classifier:
from sklearn.ensemble import RandomForestClassifier
my_RandomForest =
RandomForestClassifier(n_estimators = 19, bootstrap = True, random_state=3)
Similar to previous syntax, use my_RandomForest.fit for training your random forest
classifier and my_RandomForest.predict for prediction.

In [20]:
#initiate random forest object with given parameter
myRandomForest = RandomForestClassifier(n_estimators = 19, bootstrap = True, random_state=3)

In [21]:
#train data with random forest object
myRandomForest.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=19, n_jobs=None,
            oob_score=False, random_state=3, verbose=0, warm_start=False)

In [22]:
#predict with test data
y_predict = myRandomForest.predict(X_test)

In [23]:
#check accuracy
accuracy_log = accuracy_score(y_test,y_predict)
print(accuracy_log)

0.9245283018867925


Accuracy significantly improved using Random Forest (from 0.8301886792452831 to 0.9245283018867925)